File: packages/playground/conway-stateful/conway.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

typedef int bool;
#define true 1
#define false 0

#define ROWS 20
#define COLUMNS 20

#define A ' '
#define Z 'X'

typedef bool Cell;
typedef Cell Row[COLUMNS];
typedef Row Grid[ROWS];

typedef struct {
    int rows;
    int columns;
    Grid grid;
} Frame;

Frame createFrame(int rows, int columns) {
    Frame frame;
    frame.rows = rows;
    frame.columns = columns;
    for (int i = 0; i < rows; i++) {
        for (int j = 0; j < columns; j++) {
            frame.grid[i][j] = false;
        }
    }
    return frame;
}

Frame seedRandom(Frame frame, float chance) {
    for (int i = 0; i < frame.rows; i++) {
        for (int j = 0; j < frame.columns; j++) {
            frame.grid[i][j] = (rand() / (float)RAND_MAX < chance) ? true : false;
        }
    }
    return frame;
}

int countAliveNeighbors(Frame frame, int x, int y) {
    int count = 0;
    for (int i = -1; i <= 1; i++) {
        for (int j = -1; j <= 1; j++) {
            if (i == 0 && j == 0) {
                // skip counting ourself
                continue;
            }

            int newX = x + i;
            int newY = y + j;

            if (newX >= 0 && newX < frame.rows && newY >= 0 && newY < frame.columns && frame.grid[newX][newY]) {
                count++;
            }
        }
    }
    return count;
}

Frame updateGrid(Frame frame) {
    Frame newFrame;
    newFrame.rows = frame.rows;
    newFrame.columns = frame.columns;
    
    for (int i = 0; i < frame.rows; i++) {
        for (int j = 0; j < frame.columns; j++) {
            int aliveNeighbors = countAliveNeighbors(frame, i, j);
            if (frame.grid[i][j]) {
                // cell is currently alive
                newFrame.grid[i][j] = (aliveNeighbors == 2 || aliveNeighbors == 3) ? true : false;
            } else {
                // cell is currently dead
                newFrame.grid[i][j] = (aliveNeighbors == 3) ? true : false;
            }
        }
    }
    return newFrame;
}

// stuff for display

void clearConsole() {
    printf("\033[2J\033[H"); // ANSI escape codes to clear the screen
}

void displayGrid(Frame frame) {
    for (int i = 0; i < frame.rows; i++) {
        for (int j = 0; j < frame.columns; j++) {
            printf("%c", frame.grid[i][j] ? Z : A);
        }
        printf("\n");
    }
}

void animateGrid(Frame frame, int iterations, int delay) {
    for (int i = 0; i <= iterations; i++) {
        clearConsole();
        printf("Generation %d:\n", i);
        displayGrid(frame);
        fflush(stdout);
        usleep(delay * 1000);
        frame = updateGrid(frame);
    }
}

int main() {
    Frame frame = createFrame(ROWS, COLUMNS);
    frame = seedRandom(frame, 0.15);
    animateGrid(frame, 20, 250);
    return 0;
}

File: packages/playground/conway-stateful/conway.ts
import process from 'process';

type Cell<T = boolean> = T;
type Row<T> = Cell<T>[];
type Grid<T> = Row<T>[];

const A = ' ';
const M = '▒';
const Z = '█';

interface Frame<T> {
  rows: number;
  columns: number;
  grid: Grid<T>;
}

const countAliveNeighbors = <T>({
  rows,
  columns,
  grid,
}: Frame<T>, x: number, y: number) => {
  let count = 0;
  for (let i = -1; i <= 1; i++) {
    for (let j = -1; j <= 1; j++) {
      if (i === 0 && j === 0) {
        // Skip the current cell
        continue;
      }

      const newX = x + i;
      const newY = y + j;

      if (
        newX >= 0 &&
        newX < rows &&
        newY >= 0 &&
        newY < columns &&
        grid[newX][newY]
      ) {
        count++;
      }
    }
  }
  return count;
}

const updateGrid = (frame: Frame<boolean>) => {
  const newGrid: typeof frame.grid = [];

  for (let i = 0; i < frame.rows; i++) {
    newGrid[i] = [];
    for (let j = 0; j < frame.columns; j++) {
      const aliveNeighbors = countAliveNeighbors(frame, i, j);
      if (frame.grid[i][j]) {
        // Cell is currently alive
        newGrid[i][j] = aliveNeighbors === 2 || aliveNeighbors === 3;
      } else {
        // Cell is currently dead
        newGrid[i][j] = aliveNeighbors === 3;
      }
    }
  }
  const next: typeof frame = {
    ...frame,
    grid: newGrid,
  }
  return next;
}

const createFrame = (rows: number, columns: number) => {
  const grid: Grid<false> = Array(rows).fill(Array(columns).fill(false, 0));
  return grid;
}

const seedRandom = <T>(frame: Frame<T>, chance: number) => {
  return operate(frame, _ => Math.random() < chance);
}

const operate = <T, U>(frame: Frame<T>, operator: (cell: T) => U) => {
  const next: Frame<U> = {
    ...frame,
    grid: frame.grid.map(row => row.map(operator))
  };
  return next;
}

const displayGrid = (frame: Frame<boolean>) => {
  const strings = operate(frame, cell => cell ? Z : A);
  return strings.grid.map(columns => columns.join("")).join("\n")
}

////////////////////////////////
// Example usage:
const clearConsole = () => {
  process.stdout.write('\x1B[2J\x1B[H'); // ANSI escape codes to clear the screen
};

const animateGrid = async (
  frame: Frame<boolean>,
  iterations: number,
  delay: number,
) => {
  // recursive version instead of `for` loop
  const doAnimation = async (i: number) => {
    clearConsole();
    process.stdout.write(`Generation ${i}:\n`);
    process.stdout.write(displayGrid(frame));
    await new Promise((resolve) => setTimeout(resolve, delay));
    frame = updateGrid(frame);

    if (i < iterations) {
      doAnimation(i + 1);
    }
  }
  await doAnimation(0);
};

const main = async () => {
  const rows = 20;
  const columns = 20;
  const frame = {
    rows,
    columns,
    grid: createFrame(rows, columns),
  } as Frame<boolean>;

  let workingFrame = seedRandom(frame, 0.15);

  await animateGrid(workingFrame, 20, 250);
};

main();